### PR TITLE
feat: alter fulltext options

### LINKED ISSUE #4395: Alter column option
### What problem does the new feature solve?

since v0.9, GreptimeDB supports full-text index for string columns with the extra `fulltext` option:

```sql
CREATE TABLE `logs` (
  `message` STRING FULLTEXT,
  `time` TIMESTAMP TIME INDEX,
) WITH (
  append_mode = 'true'
);
```

But if you can't enable an existing string column `fulltext` option by alter.

### What does the feature do?

It already has the `modify` statement to change the column type:

```sql
ALTER TABLE monitor MODIFY COLUMN load_15 STRING;
```

Supports alter column options just like:

```sql
ALTER TABLE monitor SET COLUMN load_15 fulltext WITH(analyzer = 'Chinese', case_sensitive = 'false');
```

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by CookiePieWw**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

close #4395

## What's changed and what's your intention?

As in #4637, support alter fulltext options for column:
`ALTER TABLE monitor SET COLUMN <column_name> FULLTEXT WITH(analyzer = 'Chinese', case_sensitive = 'false');`

Currently we only support to change the fulltext option when:
1. Enable to disable
2. Disable to enable
3. Disable to disable (no effect actually)

Big thanks to @irenjj for his previous pr #4637 which helps a lot.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Comment by CookiePieWw** [comment]
Seems too big for a single pr... I'll try to split it  later.

File: src/sql/src/parsers/alter_parser.rs [hunk 1]
```diff
@@ -143,12 +148,42 @@ impl ParserContext<'_> {
...
+            for key in options.keys() {
+                ensure!(
+                    key == COLUMN_FULLTEXT_CHANGE_OPT_KEY_ENABLE
+                        || validate_column_fulltext_option(key),
+                    InvalidColumnOptionSnafu {
+                        name: column_name.to_string(),
+                        msg: format!("invalid FULLTEXT option: {key}"),
+                    }
+                );
+            }
```
[the remark this answers is not in the document]
Comment (CookiePieWw) [reply src/sql/src/parsers/alter_parser.rs]: The keys have been converted to lower case here actually

https://github.com/GreptimeTeam/greptimedb/blob/22f31f592930828f6bc4a40f85720dd6acd93eb4/src/sql/src/util.rs#L43-L53

File: Cargo.toml [hunk 2]
```diff
@@ -121,7 +121,7 @@ etcd-client = { version = "0.13" }
 fst = "0.4.7"
 futures = "0.3"
 futures-util = "0.3"
-greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "255f87a3318ace3f88a67f76995a0e14910983f4" }
+greptime-proto = { git = "https://github.com/CookiePieWw/greptime-proto.git", rev = "6ca31389810e1385a4887469ea96f521b2f4180c" }
```
Comment (WenyXu) [inline Cargo.toml:124]: Let's merge https://github.com/GreptimeTeam/greptime-proto/pull/196 first

File: src/store-api/src/metadata.rs [hunk 3]
```diff
@@ -749,6 +805,24 @@ pub enum MetadataError {
...
+        "Failed to set fulltext options for column {}, reason: {}",
+        column_name,
+        reason
+    ))]
+    SetFulltextOptions {
+        column_name: String,
+        reason: String,
+        #[snafu(implicit)]
+        location: Location,
+    },
```
Comment (WenyXu) [inline src/store-api/src/metadata.rs:825]:
```suggestion
    #[snafu(display("Failed to set fulltext options for column {}", column_name))]
    SetFulltextOptions {
        column_name: String,
        source: datatypes::Error,
        #[snafu(implicit)]
        location: Location,
    },
```

File: src/table/src/error.rs [hunk 4]
```diff
@@ -140,6 +140,18 @@ pub enum Error {
...
+        "Failed to set fulltext options for column {}, reason: {}",
+        column_name,
+        reason
+    ))]
+    SetFulltextOptions {
+        column_name: String,
+        reason: String,
+        #[snafu(implicit)]
+        location: Location,
+    },
```
Comment (WenyXu) [inline src/table/src/error.rs:154]: ditto

File: src/store-api/src/metadata.rs [hunk 5]
```diff
@@ -625,6 +629,58 @@ impl RegionMetadataBuilder {
...
+            if column_meta.column_schema.name == column_name {
+                ensure!(
+                    column_meta.column_schema.data_type.is_string(),
+                    SetFulltextOptionsSnafu {
+                        column_name: column_name.to_string(),
+                        reason: format!(
+                            "column {} is not a string type, but {:?}",
+                            column_name, column_meta.column_schema.data_type
+                        ),
+                    }
```
Comment (WenyXu) [inline src/store-api/src/metadata.rs:648]: let's introduce another error for this, e.g.,
```rust
#[snafu(display("Invalid column option, column name: {}, error: {}", name, msg))]
    InvalidColumnOption {
        name: String,
        msg: String,
        #[snafu(implicit)]
        location: Location,
    },
```

File: src/store-api/src/metadata.rs [hunk 6]
```diff
@@ -749,6 +805,24 @@ pub enum MetadataError {
         #[snafu(implicit)]
         location: Location,
     },
+
+    #[snafu(display("Invalid change fulltext option request"))]
+    InvalidChangeFulltextOptionRequest {
+        #[snafu(implicit)]
+        location: Location,
+    },
```
Comment (WenyXu) [inline src/store-api/src/metadata.rs:813]:
```rust
    #[snafu(display("Failed to decode proto"))]
    DecodeProto {
        #[snafu(source)]
        error: prost::DecodeError,
        #[snafu(implicit)]
        location: Location,
    },
```

File: src/store-api/src/metadata.rs [hunk 7]
```diff
@@ -639,42 +639,34 @@ impl RegionMetadataBuilder {
...
-                        reason: format!(
-                            "column {} is not a string type, but {:?}",
-                            column_name, column_meta.column_schema.data_type
-                        ),
+                    InvalidColumnOptionSnafu {
+                        column_name,
+                        msg: format!(
+                            "expect a string type, but got {}",
+                            column_meta.column_schema.data_type
+                        )
```
Comment (WenyXu) [inline src/store-api/src/metadata.rs:647]: nit: let's keep the same style
```suggestion
                        msg: "FULLTEXT index only supports string type"
```

File: Cargo.toml (hunk 2 above)
Comment (v0y4g3r) [reply Cargo.toml:124]: @CookiePieWw We can change proto commit to latest main branch now.

File: src/table/src/metadata.rs [hunk 8]
```diff
@@ -227,6 +233,80 @@ impl TableMeta {
...
+
+        ensure!(
+            !(current_fulltext_options.is_some_and(|o| o.enable) && options.enable),
+            error::InvalidColumnOptionSnafu {
+                column_name,
+                msg: "FULLTEXT index options already enabled".to_string(),
+            }
+        );
+
+        let mut columns = vec![];
```
Comment (WenyXu) [inline src/table/src/metadata.rs:273]:
```suggestion
        let mut columns = Vec::with_capacity(table_schema.column_schemas().len());
```

File: src/operator/src/expr_factory.rs [hunk 9]
```diff
@@ -488,6 +489,18 @@ pub(crate) fn to_alter_expr(
...
         }
+        AlterTableOperation::ChangeColumnFulltext {
+            column_name,
+            options,
+        } => Kind::ChangeColumnFulltext(ChangeColumnFulltext {
+            column_name: column_name.value.to_string(),
+            enable: options.enable,
+            analyzer: match options.analyzer {
+                FulltextAnalyzer::English => Analyzer::English as i32,
+                FulltextAnalyzer::Chinese => Analyzer::Chinese as i32,
```
Comment (WenyXu) [inline src/operator/src/expr_factory.rs:500]: nit:
```suggestion
                FulltextAnalyzer::English => Analyzer::English.into(),
                FulltextAnalyzer::Chinese => Analyzer::Chinese.into(),
```

File: src/mito2/src/engine/alter_test.rs [hunk 10]
```diff
@@ -438,3 +455,109 @@ async fn test_alter_on_flushing() {
...
+    // Wait for pending flush job.
+    tokio::time::timeout(Duration::from_secs(5), pending_flush_job)
+        .await
+        .unwrap()
+        .unwrap();
+    // Wait for the write job.
+    tokio::time::timeout(Duration::from_secs(5), alter_job)
+        .await
+        .unwrap()
+        .unwrap();
```
Comment (WenyXu) [inline src/mito2/src/engine/alter_test.rs:543]: nit:
```suggestion
    // Wait for the flush job.
    flush_job.await.unwrap();
    // Wait for pending flush job.
    pending_flush_job.await.unwrap();
    // Wait for the write job.
    alter_job.await.unwrap();
```

File: src/mito2/src/engine/alter_test.rs [hunk 11]
```diff
@@ -438,3 +455,109 @@ async fn test_alter_on_flushing() {
...
+        .metadata()
+        .column_by_name("tag_0")
+        .unwrap()
+        .column_schema
+        .fulltext_options()
+        .unwrap()
+        .unwrap();
+
+    assert_eq!(expect_fulltext_options, current_fulltext_options);
+    check_region_version(&engine, region_id, 1, 3, 1, 3);
```
Comment (WenyXu) [inline src/mito2/src/engine/alter_test.rs:562]: Let's verify the full text options after reopening the engine

**Review by WenyXu — APPROVED** [review]
Rest LGTM

File: src/sql/src/parsers/alter_parser.rs [hunk 12]
```diff
@@ -143,12 +148,42 @@ impl ParserContext<'_> {
...
+                .parse_options(Keyword::WITH)
+                .context(error::SyntaxSnafu)?
+                .into_iter()
+                .map(parse_option_string)
+                .collect::<Result<HashMap<String, String>>>()?;
+
+            for key in options.keys() {
+                ensure!(
+                    key.to_ascii_lowercase().as_str() == COLUMN_FULLTEXT_CHANGE_OPT_KEY_ENABLE
+                        || validate_column_fulltext_option(key.to_ascii_lowercase().as_str()),
```
Comment (fengjiachun) [inline src/sql/src/parsers/alter_parser.rs:168]: Maybe we can provide two methods：
 `validate_column_fulltext_create_option` and `validate_column_fulltext_alter_option`

```
fn validate_column_fulltext_alter_option(key) {
return key == COLUMN_FULLTEXT_CHANGE_OPT_KEY_ENABLE || validate_column_fulltext_create_option(key)
}
```

File: src/table/src/metadata.rs [hunk 13]
```diff
@@ -227,6 +233,80 @@ impl TableMeta {
...
+            .with_context(|| error::ColumnNotExistsSnafu {
+                column_name,
+                table_name,
+            })?;
+
+        ensure!(
+            column.data_type.is_string(),
+            error::InvalidColumnOptionSnafu {
+                column_name,
+                msg: "FULLTEXT index only supports string type".to_string(),
```
Comment (fengjiachun) [inline src/table/src/metadata.rs:257]:
```suggestion
                msg: "FULLTEXT index only supports string type",
```

File: src/table/src/metadata.rs [hunk 14]
```diff
@@ -227,6 +233,80 @@ impl TableMeta {
...
+
+        let current_fulltext_options = column
+            .fulltext_options()
+            .context(error::SetFulltextOptionsSnafu { column_name })?;
+
+        ensure!(
+            !(current_fulltext_options.is_some_and(|o| o.enable) && options.enable),
+            error::InvalidColumnOptionSnafu {
+                column_name,
+                msg: "FULLTEXT index options already enabled".to_string(),
```
Comment (fengjiachun) [inline src/table/src/metadata.rs:269]:
```suggestion
                msg: "FULLTEXT index options already enabled",
```

File: src/store-api/src/region_request.rs [hunk 15]
```diff
@@ -458,6 +471,32 @@ impl AlterKind {
...
+    ) -> Result<()> {
+        let column = metadata
+            .column_by_name(column_name)
+            .context(InvalidRegionRequestSnafu {
+                region_id: metadata.region_id,
+                err: format!("column {} not found", column_name),
+            })?;
+
+        ensure!(
+            column.column_schema.data_type == ConcreteDataType::string_datatype(),
```
Comment (fengjiachun) [inline src/store-api/src/region_request.rs:488]:
```suggestion
            column.column_schema.data_type.is_string(),
```

File: src/sql/src/parsers/alter_parser.rs [hunk 16]
```diff
@@ -143,12 +152,41 @@ impl ParserContext<'_> {
...
+                .parser
+                .parse_options(Keyword::WITH)
+                .context(error::SyntaxSnafu)?
+                .into_iter()
+                .map(parse_option_string)
+                .collect::<Result<HashMap<String, String>>>()?;
+
+            for key in options.keys() {
+                ensure!(
+                    validate_column_fulltext_alter_option(key),
```
Comment (fengjiachun) [inline src/sql/src/parsers/alter_parser.rs:171]: `key.to_ascii_lowercase ()` ?

Comment (CookiePieWw) [reply src/sql/src/parsers/alter_parser.rs:171]: the function `parse_option_string` has converted it to lower case already, so I removed it.

**Review by fengjiachun — APPROVED** [review]
Rest LGTM

**Pull request merged**